src/controllers/krate/search.rs

```diff
@@ -102,59 +95,70 @@ pub async fn search(app: AppState, req: Parts) -> AppResult<Json<Value>> {
                         q_string,
                     );
                     let rank = ts_rank_cd(crates::textsearchable_index_col, q);
+                    query = query.select((
+                        ALL_COLUMNS,
+                        Crate::with_name(q_string),
+                        recent_crate_downloads::downloads.nullable(),
+                        rank,
+                    ));
+                    seek = Some(Seek::Relevance);
                     query = query.then_order_by(rank.desc())
+                } else {
+                    query = query.select((
+                        ALL_COLUMNS,
+                        Crate::with_name(q_string),
+                        recent_crate_downloads::downloads.nullable(),
+                        0_f32.into_sql::<Float>(),
+                    ));
+                    seek = Some(Seek::Query);
                 }
             }
         }
 
         // Any sort other than 'relevance' (default) would ignore exact crate name matches
+        // Seek-based pagination requires a unique ordering to avoid unexpected row skipping
+        // during pagination.
+        // Therefore, when the ordering isn't unique an auxiliary ordering column should be added
+        // to ensure predictable pagination behavior.
         if sort == Some("downloads") {
-            // Custom sorting is not supported yet with seek.
-            supports_seek = false;
-
-            query = query.order(crates::downloads.desc())
+            seek = Some(Seek::Downloads);
+            query = query.order((crates::downloads.desc(), crates::id.asc()))
         } else if sort == Some("recent-downloads") {
-            // Custom sorting is not supported yet with seek.
-            supports_seek = false;
-
-            query = query.order(recent_crate_downloads::downloads.desc().nulls_last())
+            seek = Some(Seek::RecentDownloads);
+            query = query.order((
+                recent_crate_downloads::downloads.desc().nulls_last(),
+                crates::id.asc(),
+            ))
         } else if sort == Some("recent-updates") {
-            // Custom sorting is not supported yet with seek.
-            supports_seek = false;
-
-            query = query.order(crates::updated_at.desc());
+            seek = Some(Seek::RecentUpdates);
+            query = query.order((crates::updated_at.desc(), crates::id.asc()));
         } else if sort == Some("new") {
-            // Custom sorting is not supported yet with seek.
-            supports_seek = false;
-
-            query = query.order(crates::created_at.desc());
+            seek = Some(Seek::New);
+            query = query.order((crates::created_at.desc(), crates::id.asc()));
```

Comment:
If larger `ids` signify newer items with the same `created_at` timestamp, it might be more intuitive to switch the sorting order from ascending `crates::id.asc()` to descending `crates::id.desc()`.

Reply:
> If larger `ids` signify newer items with the same `created_at` timestamp

yep, that's how it would be. I'm not sure if it's worth breaking the consistency. I haven't checked, but I would be surprised if we actually had crates in the database with the exact same `created_at` timestamp.

Reply:
If we can guarantee that timestamps will never be identical, then the auxiliary ordering column becomes unnecessary and can be removed. However, if the possibility of duplicate timestamps exists, then it's better to keep it.

---

Since we all use descending order for sorting, I think it would be more intuitive to change all the auxiliary column `id.asc()` to `id.desc()`. What do you think?

Reply:
> I think it would be more intuitive to change all the auxiliary column `id.asc()` to `id.desc()`

yeah, that should also work :)

Reply:
It is implemented in the new commit.